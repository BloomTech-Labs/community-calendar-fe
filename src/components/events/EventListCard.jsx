import React from 'react'
import moment from 'moment'
import {Link} from 'react-router-dom'
import PropTypes from 'prop-types'

//styles
import {
  column,
  columns,
  isNarrow,
  listCard,
  gridCard,
  eventDescription,
  event_details,
  space_letters,
  event_image,
  descriptionUnderline,
  title,
  description,
  tileTitle,
  dont_break_out,
  descriptionFade,
} from './styles/EventListCard.module.scss'

import {TimeIcon, CalendarIcon} from '../icons'

/* 
The cards shown on the EventList component.
Changing the class names on the Link element
sets the style to List format or Grid format
 */
export default function EventListCard(props) {
  const {item, useListView} = props
  const location =
    item.locations && item.locations[0].streetAddress
      ? item.locations[item.locations.length - 1]
      : null

  return (
    //  Use `listCard` or `gridCard` style based on `useListView` bool from parent
    <Link
      className={useListView ? `${listCard}  ${columns}` : gridCard}
      to={`/events/${item.id}`}
    >
      {item.eventImages && item.eventImages.length > 0 && (
        <img
          src={item.eventImages[0].url}
          alt={item.title}
          className={(column, isNarrow, event_image)}
        />
      )}
      <div className={`${column} ${event_details}`}>
        <div
          data-id='event_location'
          data-testid='event_location'
          className='has-text-weight-bold color_chalice is-flex justify-between'
        >
          {/* display neighborhood if defined, otherwise city */}
          <p className='is-size-7-mobile is-size-6-tablet '>
            <span className='is-uppercase'>
              {location.neighborhood ? location.neighborhood : location.city}
            </span>
            {/* display distanceFromUser if defined */}
            {location && location.distanceFromUser && (
              <>
                <span className='color_shark is-size-7 '>
                  &nbsp; &#8226; &nbsp;
                </span>
                <span
                  className={`${space_letters} is-size-7 has-text-weight-normal`}
                >
                  {`${location.distanceFromUser.toFixed(1)}`}
                  &nbsp;
                  {`${location.distanceUnit === 'miles' ? 'mi' : 'km'}`}
                </span>
                <span className='hidden-xs has-text-weight-normal is-size-7'>
                  &nbsp;away
                </span>
              </>
            )}
          </p>
          {Math.ceil(
            moment.duration(moment(item.end).diff(moment(item.start))).asDays(),
          ) === 1 && (
            <p className='is-hidden-tablet is-size-6-tablet is-size-7-mobile is-inline-block'>
              <span data-id='event_time_mobile' className='color_chalice'>
                {`${moment(item.start).format('ll')}`}
              </span>
            </p>
          )}
        </div>
        <p
          data-id='event_title'
          data-testid='event_title'
          className={`is-size-5-tablet is-size-6-mobile has-text-weight-bold color_black ${title} ${
            !useListView ? tileTitle : ''
          }`}
        >
          {item.title}
        </p>
        {Math.ceil(
          moment.duration(moment(item.end).diff(moment(item.start))).asDays(),
        ) === 1 && (
          <p className='is-hidden-mobile is-size-6-tablet is-size-7-mobile'>
            <span data-id='event_time' className='color_chalice'>
              {`${moment(item.start).format('ddd, MMMM Do YYYY')}`}
            </span>
          </p>
        )}
        <p className='is-size-6-tablet is-size-7-mobile '>
          <span data-id='event_time' className='color_chalice hidden-xs'>
            {Math.ceil(
              moment
                .duration(moment(item.end).diff(moment(item.start)))
                .asDays(),
            ) === 1
              ? `${moment(item.start).format('h:mm a')} - ${moment(
                  item.end,
                ).format('h:mm a')}`
              : `${moment(item.start).format('MMM Do YYYY h:mm a')} - ${moment(
                  item.end,
                ).format('MMM Do YYYY h:mm a')}`}
          </span>
          <span className='color_chalice hidden-xs'>
            <span className='color_shark is-size-6-tablet hidden-xs'>
              &nbsp;&#8226;&nbsp;
            </span>
            {item.ticketPrice ? `$${item.ticketPrice}` : 'FREE'}
          </span>
        </p>
        <p
          data-id='event_description'
          data-testid='event_description'
          className={`${eventDescription} ${description} is-size-7 color_black is-hidden-mobile`}
        >
          {item.description}
        </p>
        <div
          className={`${descriptionFade} ${useListView ? '' : 'd-none'}`}
        ></div>
        <div className={descriptionUnderline}></div>
      </div>
    </Link>
  )
}

EventListCard.propTypes = {
  item: PropTypes.object,
  useListView: PropTypes.bool,
}
