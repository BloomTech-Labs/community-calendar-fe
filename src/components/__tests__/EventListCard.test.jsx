import React from 'react'
import {render} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import userEvent from '@testing-library/user-event'
import EventListCard from '../events/EventListCard'
import {MemoryRouter} from 'react-router-dom'
import moment from 'moment'

//mock data
import {
  TEST_FULL,
  TEST_NO_NEIGHBORHOOD,
  TEST_NO_DISTANCE,
} from '../../../test-utils/mock_data/test_event'

jest.mock('moment', () => {
  const moment = require.requireActual('moment')
  // moment is both an object and function so create an actual instance of it
  const momentInstance = moment()

  // spy on methods of moment and modify them
  jest.spyOn(momentInstance, 'format').mockImplementation(() => '')

  // create a fake instance of moment
  function fakeMoment() {
    return momentInstance
  }

  // apply properties of moment to fakeMoment
  Object.assign(fakeMoment, moment)
  return fakeMoment
})

describe('Tests for EventListCard', () => {
  test('TEST_FULL render should match the snapshot', () => {
    const tree = render(
      <MemoryRouter>
        <EventListCard item={TEST_FULL} />
      </MemoryRouter>,
    )
    expect(tree).toMatchSnapshot()
  })

  test('Should render TEST_FULL data from props', () => {
    const {getByText, getByTestId} = render(
      <MemoryRouter>
        <EventListCard item={TEST_FULL} />
      </MemoryRouter>,
    )

    expect(getByTestId('event_description')).toHaveTextContent(
      TEST_FULL.description,
    )
    expect(getByTestId('event_title')).toHaveTextContent(TEST_FULL.title)

    // only neighborhood, not city, displays if neighborhood provided
    expect(getByTestId('event_location')).toHaveTextContent(
      TEST_FULL.locations[0].neighborhood,
    )
    expect(getByTestId('event_location')).not.toHaveTextContent(
      TEST_FULL.locations[0].city,
    )

    // distance from user displays when provided
    expect(getByTestId('event_location')).toHaveTextContent(
      TEST_FULL.locations[0].distanceFromUser.toFixed(1),
    )
    expect(getByTestId('event_location')).toHaveTextContent(/mi away/i)
  })

  test('Should render TEST_NO_NEIGHBORHOOD data from props', () => {
    const {getByTestId} = render(
      <MemoryRouter>
        <EventListCard item={TEST_NO_NEIGHBORHOOD} />
      </MemoryRouter>,
    )

    // only city displays if no neighborhood provided
    expect(getByTestId('event_location')).not.toHaveTextContent(
      TEST_NO_NEIGHBORHOOD.locations[0].neighborhood,
    )
    expect(getByTestId('event_location')).toHaveTextContent(
      TEST_NO_NEIGHBORHOOD.locations[0].city,
    )
  })

  test('Should render TEST_NO_DISTANCE data from props', () => {
    const {getByTestId} = render(
      <MemoryRouter>
        <EventListCard item={TEST_NO_DISTANCE} />
      </MemoryRouter>,
    )

    // distance from user does not display when not provided
    expect(getByTestId('event_location')).not.toHaveTextContent(/mi away/i)
  })
})
