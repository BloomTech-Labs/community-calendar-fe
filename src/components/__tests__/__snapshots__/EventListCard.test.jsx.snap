// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for EventListCard TEST_FULL render should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        class="gridCard"
        href="/events/ck5flbejv00200730x34y4le6"
      >
        <img
          alt="Art & Architecture - Downtown Walking Tour"
          class="event_image"
          src="https://img.evbuc.com/https%3A%2F%2Fcdn.evbuc.com%2Fimages%2F69917729%2F89373299245%2F1%2Foriginal.jpg?w=800&auto=format%2Ccompress&q=75&sharp=10&rect=0%2C693%2C4000%2C2000&s=94184073d91c25306d22d1b0401cfea9"
        />
        <div
          class="column event_details"
        >
          <div
            class="has-text-weight-bold color_chalice is-flex justify-between"
            data-id="event_location"
            data-testid="event_location"
          >
            <p
              class="is-size-7-mobile is-size-6-tablet "
            >
              <span
                class="is-uppercase"
              >
                Downtown
              </span>
              <span
                class="color_shark is-size-7 "
              >
                  •  
              </span>
              <span
                class="space_letters is-size-7 has-text-weight-normal"
              >
                9402.3
                 
                mi
              </span>
              <span
                class="hidden-xs has-text-weight-normal is-size-7"
              >
                 away
              </span>
            </p>
          </div>
          <p
            class="is-size-5-tablet is-size-6-mobile has-text-weight-bold color_black title tileTitle"
            data-id="event_title"
            data-testid="event_title"
          >
            Art & Architecture - Downtown Walking Tour
          </p>
          <p
            class="is-size-6-tablet is-size-7-mobile "
          >
            <span
              class="color_chalice hidden-xs"
              data-id="event_time"
            >
               - 
            </span>
            <span
              class="color_chalice hidden-xs"
            >
              <span
                class="color_shark is-size-6-tablet hidden-xs"
              >
                 • 
              </span>
              $FREE
            </span>
          </p>
          <p
            class="eventDescription description is-size-7 color_black is-hidden-mobile"
            data-id="event_description"
            data-testid="event_description"
          >
            Detroit is a city rich in history, grand buildings, and vibrant art surrounding public spaces. Walk with Detroit Experience Factory through the downtown area as we explore some of the great contributions of both prolific architects and emerging artists.
          </p>
          <div
            class="descriptionFade d-none"
          />
          <div
            class="descriptionUnderline"
          />
        </div>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="gridCard"
      href="/events/ck5flbejv00200730x34y4le6"
    >
      <img
        alt="Art & Architecture - Downtown Walking Tour"
        class="event_image"
        src="https://img.evbuc.com/https%3A%2F%2Fcdn.evbuc.com%2Fimages%2F69917729%2F89373299245%2F1%2Foriginal.jpg?w=800&auto=format%2Ccompress&q=75&sharp=10&rect=0%2C693%2C4000%2C2000&s=94184073d91c25306d22d1b0401cfea9"
      />
      <div
        class="column event_details"
      >
        <div
          class="has-text-weight-bold color_chalice is-flex justify-between"
          data-id="event_location"
          data-testid="event_location"
        >
          <p
            class="is-size-7-mobile is-size-6-tablet "
          >
            <span
              class="is-uppercase"
            >
              Downtown
            </span>
            <span
              class="color_shark is-size-7 "
            >
                •  
            </span>
            <span
              class="space_letters is-size-7 has-text-weight-normal"
            >
              9402.3
               
              mi
            </span>
            <span
              class="hidden-xs has-text-weight-normal is-size-7"
            >
               away
            </span>
          </p>
        </div>
        <p
          class="is-size-5-tablet is-size-6-mobile has-text-weight-bold color_black title tileTitle"
          data-id="event_title"
          data-testid="event_title"
        >
          Art & Architecture - Downtown Walking Tour
        </p>
        <p
          class="is-size-6-tablet is-size-7-mobile "
        >
          <span
            class="color_chalice hidden-xs"
            data-id="event_time"
          >
             - 
          </span>
          <span
            class="color_chalice hidden-xs"
          >
            <span
              class="color_shark is-size-6-tablet hidden-xs"
            >
               • 
            </span>
            $FREE
          </span>
        </p>
        <p
          class="eventDescription description is-size-7 color_black is-hidden-mobile"
          data-id="event_description"
          data-testid="event_description"
        >
          Detroit is a city rich in history, grand buildings, and vibrant art surrounding public spaces. Walk with Detroit Experience Factory through the downtown area as we explore some of the great contributions of both prolific architects and emerging artists.
        </p>
        <div
          class="descriptionFade d-none"
        />
        <div
          class="descriptionUnderline"
        />
      </div>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
